apple/DemoApp/Demo/DemoNavigationView.swift

```diff
@@ -16,10 +16,11 @@ struct DemoNavigationView: View {
     private let navigationDelegate = NavigationDelegate()
     // NOTE: This is probably not ideal but works for demo purposes.
     // This causes a thread performance checker warning log.
-    private let spokenInstructionObserver = AVSpeechSpokenInstructionObserver(isMuted: false)
+    @State private var spokenInstructionObserver = AVSpeechSpokenInstructionObserver(isMuted: false)
```

Comment:
This doesn't look quite right, but it's probably a result of our `AVSpeechSpokenInstructionObserver`. That class doesn't really have a good way to publish the isMuted value. 

Here's an updated version which you could copy & replace to the `apple/Sources/FerrostarCore/Speech.swift` file:

```swift

public protocol SpokenInstructionObserver {
    
    /// Mute or unmute the speech engine.
    ///
    /// - Parameter mute: Mute the speech engine if true, unmute if false.
    func setMute(_ mute: Bool)
    
    /// Handles spoken instructions as they are triggered.
    ///
    /// As long as it is used with the supplied ``FerrostarCore`` class,
    /// implementors may assume this function will never be called twice
    /// for the same instruction during a navigation session.
    func spokenInstructionTriggered(_ instruction: SpokenInstruction)

    /// Stops speech and clears the queue of spoken utterances.
    func stopAndClearQueue()

    /// If the speech synthisizer is currently muted.
    var isMuted: Bool { get }
}

public class AVSpeechSpokenInstructionObserver: ObservableObject, SpokenInstructionObserver {
    
    @Published public private(set) var isMuted: Bool
    
    private let synthesizer = AVSpeechSynthesizer()

    public init(isMuted: Bool) {
        self.isMuted = isMuted
    }

    public func setMute(_ mute: Bool) {
        guard isMuted != mute else {
            return
        }
        
        // Immediately set the publisher. This will avoid delays updating UI.
        isMuted = mute
        
        if mute && synthesizer.isSpeaking {
            synthesizer.stopSpeaking(at: .immediate)
        }
    }
    
    public func spokenInstructionTriggered(_ instruction: FerrostarCoreFFI.SpokenInstruction) {
        guard !isMuted else {
            return
        }

        let utterance: AVSpeechUtterance = if #available(iOS 16.0, *),
                                              let ssml = instruction.ssml,
                                              let ssmlUtterance = AVSpeechUtterance(ssmlRepresentation: ssml)
        {
            ssmlUtterance
        } else {
            AVSpeechUtterance(string: instruction.text)
        }

        synthesizer.speak(utterance)
    }

    public func stopAndClearQueue() {
        synthesizer.stopSpeaking(at: .immediate)
    }
}
```

I'd be happy to also post a PR with this change if that'd help. 

The key difference is the `AVSpeechSpokenInstructionObserver` will actually have an isMuted publisher on it that's designed to be used in a SwiftUI instead of a simple bool. A simple bool won't trigger updates to the button, causing the problem you were probably hoping to solve with `@State`.

Reply:
This is tricky. You're right that the current protocol isn't going to just drop in to a SwiftUI setup.... but I'm not sure this actually fixes the problem 😅 It handles one half (publishing updates), but it doesn't actually get us to a two-way binding.

Reply:
In a SwiftUI app, I would not expect a `@Binding var` to actually alter the behavior of a service even though you're correct that it offers a two way street which could be used for that. Having a bespoke `func setMute(_ mute: Bool)` and a published state that SwiftUI can easily use seems to capture the best of a button running a behavior and a state being published for the visual UI.

A get set pattern was nice, but in this case seems like the wrong solution given the goals and SwiftUI.

Reply:
Yeah, I see your point. Though I think `@Binding` is at least somewhat intended for this purpose.

The problem I hit with the method-based approach, which is how you'd do it in Jetpack Compose to preserve UDF, is that things get _extremely_ messy with the Swift type system.


At A high level, here's what I think you're suggesting:

`SpokenInstructionObserver`:
* Make `isMuted` get-only
* Add `func setMuted(...)`

`MuteUIButton`:
* Switch from `@Binding` to `@ObservedObject` and observe the `SpokenInstructionObserver` directly
* Rather than toggling the binding, invoke `spokenInstructionObserver.setMuted(...)`
* Since `spokenInstructionObserver.isMuted` is published, observe that value in the image composition rather than the binding

In order to use a `@Published` property and observe the object directly, we would need to make `SpokenInstructionObserver` conform to `ObservableObject`. This sounds trivial, but this opens up Pandora's box with errors like `Type 'any SpokenInstructionObserver' cannot conform to 'ObservableObject'` that highlight the flaws of the Swift type system...

I am unable to find a resolution to this at the moment. Any ideas?

Reply:
Okay... after banging my head against the wall a bit more, I think I have something that works. It's nominally better... Have a look over it and lemme know what you think. I don't really like the way it forces us to include generic bounds. And I couldn't get it to work with an optional `T` in the constructor, since the desirable default, `nil` is going to have an unknown type 😂 And `any` types don't work as noted. So, the only option remaining that seems to work OK is to define a dummy default, which you can override.